ssl/event_queue.c

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include "internal/event_queue.h"
+#include "crypto/sparse_array.h"
+#include "ssl_local.h"
+
+struct ossl_event_subscription_s {
+    struct ossl_event_subscription_s *prev, *next;
+    ossl_event_callback_fn *cb;
+    void *cbarg;
+    OSSL_EVENT_QUEUE *queue;
+    uint32_t type;
+};
+
+DEFINE_SPARSE_ARRAY_OF(OSSL_EVENT_SUBSCRIPTION);
+
+struct ossl_event_queue_s {
+    PRIORITY_QUEUE_OF(OSSL_EVENT) *events;
+    SPARSE_ARRAY_OF(OSSL_EVENT_SUBSCRIPTION) *subscribers;
+};
+
+static int event_compare(const OSSL_EVENT *a, const OSSL_EVENT *b)
+{
+    int i = ossl_time_compare(a->when, b->when);
+
+    if (i != 0)
+        return i;
+    if (a->priority > b->priority)
+        return 1;
+    if (a->priority < b->priority)
+        return -1;
+    return 0;
+}
+
+OSSL_EVENT_QUEUE *ossl_event_queue_new(void)
+{
+    OSSL_EVENT_QUEUE *r = OPENSSL_malloc(sizeof(*r));
+
+    if (r != NULL) {
+        r->events = ossl_pqueue_OSSL_EVENT_new(&event_compare);
+        r->subscribers = ossl_sa_OSSL_EVENT_SUBSCRIPTION_new();
+        if (r->events == NULL || r->subscribers == NULL) {
+            ossl_event_queue_free(r);
+            return NULL;
+        }
+    }
+    return r;
+}
+
+void ossl_event_queue_free(OSSL_EVENT_QUEUE *queue)
+{
+    OSSL_EVENT *e;
+
+    /* This isn't remotely correct yet */
+    if (queue != NULL) {
+        if (queue->events != NULL) {
+            while ((e = ossl_pqueue_OSSL_EVENT_pop(queue->events)) != NULL)
+                OPENSSL_free(e);
+            ossl_pqueue_OSSL_EVENT_free(queue->events);
+        }
+        ossl_sa_OSSL_EVENT_SUBSCRIPTION_free(queue->subscribers);
+        OPENSSL_free(queue);
+    }
+}
+
+OSSL_EVENT *ossl_event_new(uint32_t type, uint32_t priority,
+                           OSSL_TIME when, void *ctx,
+                           void *payload, size_t payload_size)
+{
+    OSSL_EVENT *e = OPENSSL_malloc(sizeof(*e));
+
+    if (e != NULL) {
+        ossl_event_set(e, type, priority, when, ctx, payload, payload_size);
+        e->flag_dynamic = 1;
```

Comment:
This makes me wonder why we need non-dynamic events.

Reply:
A non-dynamic event can be embedded into another structure.  i.e. cuts overhead allocating and freeing.
E.g. a re-transmit timer event will always exist with a connection -- do we need to allocate it separately?

Dropping non-dynamic is possible, but it feels like it will be useful.